File: src/main/java/com/app/core/ui/view/UITextView.java
package com.app.core.ui.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.app.core.action.Action1;
import com.app.core.action.Action2;
import com.app.core.ui.layout.UILayout;

public final class UITextView extends UITextStyle{
    /* For TexView Construct*/
    public static TextView create(@NonNull Context context, Action1<TextView> action) {
        TextView text = new TextView(context);
        if (action != null) action.invoke(text);
        return text;
    }

    public static TextView create(@NonNull ViewGroup parent, Action1<TextView> action) {
        return create(parent.getContext(), action);
    }

    public static TextView create(@NonNull RelativeLayout parent, Action2<TextView, RelativeLayout.LayoutParams> action, int width, int height) {
        TextView text = new TextView(parent.getContext());
        RelativeLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        return text;
    }

    public static TextView create(@NonNull RelativeLayout parent, Action2<TextView, RelativeLayout.LayoutParams> action) {
        return create(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static TextView create(@NonNull LinearLayout parent, Action2<TextView, LinearLayout.LayoutParams> action, int width, int height) {
        TextView text = new TextView(parent.getContext());
        LinearLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        return text;
    }

    public static TextView create(@NonNull LinearLayout parent, Action2<TextView, LinearLayout.LayoutParams> action) {
        return create(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static TextView create(@NonNull FrameLayout parent, Action2<TextView, FrameLayout.LayoutParams> action, int width, int height) {
        TextView text = new TextView(parent.getContext());
        FrameLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        return text;
    }

    public static TextView create(@NonNull FrameLayout parent, Action2<TextView, FrameLayout.LayoutParams> action) {
        return create(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static TextView add(@NonNull ViewGroup parent, Action1<TextView> action) {
        TextView text = new TextView(parent.getContext());
        if (action != null) action.invoke(text);
        parent.addView(text);
        return text;
    }

    public static TextView add(@NonNull RelativeLayout parent, Action2<TextView, RelativeLayout.LayoutParams> action, int width, int height) {
        TextView text = new TextView(parent.getContext());
        RelativeLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        parent.addView(text);
        return text;
    }

    public static TextView add(@NonNull RelativeLayout parent, Action2<TextView, RelativeLayout.LayoutParams> action) {
        return add(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static TextView add(@NonNull LinearLayout parent, Action2<TextView, LinearLayout.LayoutParams> action, int width, int height) {
        TextView text = new TextView(parent.getContext());
        LinearLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        parent.addView(text);
        return text;
    }

    public static TextView add(@NonNull LinearLayout parent, Action2<TextView, LinearLayout.LayoutParams> action) {
        return add(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static TextView add(@NonNull FrameLayout parent, Action2<TextView, FrameLayout.LayoutParams> action, int width, int height) {
        TextView text = new TextView(parent.getContext());
        FrameLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        parent.addView(text);
        return text;
    }

    public static TextView add(@NonNull FrameLayout parent, Action2<TextView, FrameLayout.LayoutParams> action) {
        return add(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    /* For EditText construct*/
    public static EditText createEdit(@NonNull Context context, Action1<EditText> action) {
        EditText text = new EditText(context);
        if (action != null) action.invoke(text);
        return text;
    }

    public static EditText createEdit(@NonNull ViewGroup parent, Action1<EditText> action) {
        return createEdit(parent.getContext(), action);
    }

    public static EditText createEdit(@NonNull RelativeLayout parent, Action2<EditText, RelativeLayout.LayoutParams> action, int width, int height) {
        EditText text = new EditText(parent.getContext());
        RelativeLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        return text;
    }

    public static EditText createEdit(@NonNull RelativeLayout parent, Action2<EditText, RelativeLayout.LayoutParams> action) {
        return createEdit(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static EditText createEdit(@NonNull LinearLayout parent, Action2<EditText, LinearLayout.LayoutParams> action, int width, int height) {
        EditText text = new EditText(parent.getContext());
        LinearLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        return text;
    }

    public static EditText createEdit(@NonNull LinearLayout parent, Action2<EditText, LinearLayout.LayoutParams> action) {
        return createEdit(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static EditText createEdit(@NonNull FrameLayout parent, Action2<EditText, FrameLayout.LayoutParams> action, int width, int height) {
        EditText text = new EditText(parent.getContext());
        FrameLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        return text;
    }

    public static EditText createEdit(@NonNull FrameLayout parent, Action2<EditText, FrameLayout.LayoutParams> action) {
        return createEdit(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static EditText addEdit(@NonNull ViewGroup parent, Action1<EditText> action) {
        return createEdit(parent.getContext(), action);
    }

    public static EditText addEdit(@NonNull RelativeLayout parent, Action2<EditText, RelativeLayout.LayoutParams> action, int width, int height) {
        EditText text = new EditText(parent.getContext());
        RelativeLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        parent.addView(text);
        return text;
    }

    public static EditText addEdit(@NonNull RelativeLayout parent, Action2<EditText, RelativeLayout.LayoutParams> action) {
        return addEdit(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static EditText addEdit(@NonNull LinearLayout parent, Action2<EditText, LinearLayout.LayoutParams> action, int width, int height) {
        EditText text = new EditText(parent.getContext());
        LinearLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        parent.addView(text);
        return text;
    }

    public static EditText addEdit(@NonNull LinearLayout parent, Action2<EditText, LinearLayout.LayoutParams> action) {
        return addEdit(parent, action, UILayout.Wrap, UILayout.Wrap);
    }

    public static EditText addEdit(@NonNull FrameLayout parent, Action2<EditText, FrameLayout.LayoutParams> action, int width, int height) {
        EditText text = new EditText(parent.getContext());
        FrameLayout.LayoutParams params = UILayout.wrap(parent, width, height);
        text.setLayoutParams(params);
        if (action != null) action.invoke(text, params);
        parent.addView(text);
        return text;
    }

    public static EditText addEdit(@NonNull FrameLayout parent, Action2<EditText, FrameLayout.LayoutParams> action) {
        return addEdit(parent, action, UILayout.Wrap, UILayout.Wrap);
    }
}

File: src/main/java/com/app/core/ui/layout/UITextView.kt
package com.app.core.ui.layout

import android.app.Activity
import android.app.Dialog
import android.support.constraint.ConstraintLayout
import android.support.design.widget.CollapsingToolbarLayout
import android.support.design.widget.CoordinatorLayout
import android.support.v4.app.Fragment
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import android.widget.*

fun <T : Activity> T.textView(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : Fragment> T.textView(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : Dialog> T.textView(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : ViewGroup> T.textView(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun RelativeLayout.textView(
    init: (TextView.(RelativeLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun LinearLayout.textView(
    init: (TextView.(LinearLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun FrameLayout.textView(
    init: (TextView.(FrameLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun ConstraintLayout.textView(
    init: (TextView.(ConstraintLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun CoordinatorLayout.textView(
    init: (TextView.(CoordinatorLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun CollapsingToolbarLayout.textView(
    init: (TextView.(CollapsingToolbarLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun RecyclerView.textView(
    init: (TextView.(RecyclerView.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): TextView {
    val view = TextView(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun <T : Activity> T.editText(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Match,
    height: Int = Match
): EditText {
    val view = EditText(this)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : Fragment> T.editText(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Match,
    height: Int = Match
): EditText {
    val view = EditText(this.context)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : Dialog> T.editText(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Match,
    height: Int = Match
): EditText {
    val view = EditText(this.context)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : ViewGroup> T.editText(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun RelativeLayout.editText(
    init: (TextView.(RelativeLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun LinearLayout.editText(
    init: (TextView.(LinearLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun FrameLayout.editText(
    init: (TextView.(FrameLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun ConstraintLayout.editText(
    init: (TextView.(ConstraintLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun CoordinatorLayout.editText(
    init: (TextView.(CoordinatorLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun CollapsingToolbarLayout.editText(
    init: (TextView.(CollapsingToolbarLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun RecyclerView.editText(
    init: (TextView.(RecyclerView.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): EditText {
    val view = EditText(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun <T : Activity> T.button(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : Fragment> T.button(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : Dialog> T.button(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    view.layoutParams = params(width, height)
    init?.let { view.it(view.layoutParams) }
    return view
}

fun <T : ViewGroup> T.button(
    init: (TextView.(ViewGroup.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun RelativeLayout.button(
    init: (TextView.(RelativeLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun LinearLayout.button(
    init: (TextView.(LinearLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun FrameLayout.button(
    init: (TextView.(FrameLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun ConstraintLayout.button(
    init: (TextView.(ConstraintLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun CoordinatorLayout.button(
    init: (TextView.(CoordinatorLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun CollapsingToolbarLayout.button(
    init: (TextView.(CollapsingToolbarLayout.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}

fun RecyclerView.button(
    init: (Button.(RecyclerView.LayoutParams) -> Unit)? = null,
    width: Int = Wrap,
    height: Int = Wrap
): Button {
    val view = Button(this.context)
    val params = params(width, height)
    view.layoutParams = params
    init?.let { view.it(params) }
    addView(view)
    return view
}